pkg/deploy/gateway.go

```diff
@@ -0,0 +1,494 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// GatewayServiceName is the name of the service which through which the gateway can be accessed
+	GatewayServiceName = "che-gateway"
+
+	gatewayServerConfigName    = "che-gateway-route-server"
+	gatewayConfigComponentName = "che-gateway-config"
+)
+
+var (
+	serviceAccountDiffOpts = cmpopts.IgnoreFields(corev1.ServiceAccount{}, "TypeMeta", "ObjectMeta", "Secrets", "ImagePullSecrets")
+	roleDiffOpts           = cmpopts.IgnoreFields(rbac.Role{}, "TypeMeta", "ObjectMeta")
+	roleBindingDiffOpts    = cmpopts.IgnoreFields(rbac.RoleBinding{}, "TypeMeta", "ObjectMeta")
+	serviceDiffOpts        = cmp.Options{
+		cmpopts.IgnoreFields(corev1.Service{}, "TypeMeta", "ObjectMeta", "Status"),
+		cmpopts.IgnoreFields(corev1.ServiceSpec{}, "ClusterIP"),
+	}
+	configMapDiffOpts = cmpopts.IgnoreFields(corev1.ConfigMap{}, "TypeMeta", "ObjectMeta")
+)
+
+// SyncGatewayToCluster installs or deletes the gateway based on the custom resource configuration
+func SyncGatewayToCluster(instance *orgv1.CheCluster, clusterAPI ClusterAPI) error {
+	if instance.Spec.Server.ServerExposureStrategy == "single-host" &&
+		(GetSingleHostExposureType(instance) == "gateway") {
+		return syncAll(instance, clusterAPI)
+	}
+
+	return deleteAll(instance, clusterAPI)
+}
+
+func syncAll(instance *orgv1.CheCluster, clusterAPI ClusterAPI) error {
+	sa := getGatewayServiceAccountSpec(instance)
+	if err := sync(instance, clusterAPI, &sa, serviceAccountDiffOpts); err != nil {
+		return err
+	}
+
+	role := getGatewayRoleSpec(instance)
+	if err := sync(instance, clusterAPI, &role, roleDiffOpts); err != nil {
+		return err
+	}
+
+	roleBinding := getGatewayRoleBindingSpec(instance)
+	if err := sync(instance, clusterAPI, &roleBinding, roleBindingDiffOpts); err != nil {
+		return err
+	}
+
+	traefikConfig := getGatewayTraefikConfigSpec(instance)
+	if err := sync(instance, clusterAPI, &traefikConfig, configMapDiffOpts); err != nil {
+		return err
+	}
+
+	depl := getGatewayDeploymentSpec(instance)
+	if err := sync(instance, clusterAPI, &depl, deploymentDiffOpts); err != nil {
+		return err
+	}
+
+	service := getGatewayServiceSpec(instance)
+	if err := sync(instance, clusterAPI, &service, serviceDiffOpts); err != nil {
+		return err
+	}
+
+	serverConfig := getGatewayServerConfigSpec(instance)
+	if err := sync(instance, clusterAPI, &serverConfig, configMapDiffOpts); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deleteAll(instance *orgv1.CheCluster, clusterAPI ClusterAPI) error {
+	deployment := appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GatewayServiceName,
+			Namespace: instance.Namespace,
+		},
+	}
+	if err := delete(clusterAPI, &deployment); err != nil {
```

Comment:
if it is NotFound error, should be ignore it instead returning?

Reply:
`delete()` does a "fetch&delete" (which is an approach used on other places in the codebase, too), so a chance of that happening are IMHO low, but it is a good point nonetheless.